src/Storages/StorageS3.h

```diff
@@ -83,6 +83,7 @@ class StorageS3Source : public SourceWithKeyCondition, WithContext
 
         KeyWithInfoPtr next(size_t idx = 0) override; /// NOLINT
         size_t estimatedKeysCount() override;
+        bool hasMore();
```

Comment:
Maybe get rid of `hasMore()`, and modify `IITerator::estimatedKeysCount()` to return `int` (instead of `size_t`), with following semantics:
- `< 0` - means that estimation wasn't possible, there are potentially many keys
- '> 0' - there are at least this number of keys in the bucket matching filtering criteria.

This would clear implementation a little bit and prevent bleeding of knowledge about `StorageS3Source::DisclosedGlobIterator` (and its special protocols) into other parts of the code.

It looks like `int` is going to be enough, that provides estimation of up to `2*31` keys. Anything more overflows into the `< 0` case, which means (not unable to estimate, probably a lot of keys). And it looks like current implementation gives estimation of 1000 keys max.

Reply:
or you can have `constexpr size_t UNABLE_TO_ESTIMATE = std::numeric_limits<size_t>::max();`  and treat that as a special value, without changing the signature.

Reply:
This also sounds fine. The returned value should not exceed 1000 anyway